### PR TITLE
Add `updateEntity` BP hook

### PR DESCRIPTION
**Pull request by thehabbos007**
## 🌟 What is the purpose of this PR?

This PR adds an `updateEntity` hook to use for the new graph type system. 

@yusufkinatas example usage:  https://github.com/hashintel/hash/commit/d41b4e141531e6accbfd3e443d565ced3622e51c

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1202953520344821/1203086812814270/f) _(internal)_

## 🚫 Blocked by

- Perhaps merging?

## 🔍 What does this change?

- This adds a new GQL mutation, `updateEntity`
- Adds a new BP shim message `UpdateEntityRequest`
- Adds a new BP hook ``useBlockProtocolUpdateEntity`
- Exposes the new `updateEntity` functionality in `useBlockProtocolFunctionsWithOntology`

## 🛡 What tests cover this?

- Existing integration tests on the backend

## ❓ How to test this/DEMO

Here is an example usage
https://github.com/hashintel/hash/commit/d41b4e141531e6accbfd3e443d565ced3622e51c

### CODE REVIEW

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/ontology/useBlockProtocolGetDataType.ts [hunk 1]
```diff
@@ -14,8 +14,8 @@ export const useBlockProtocolGetDataType = (): {
   const [getFn] = useLazyQuery<GetDataTypeQuery, GetDataTypeQueryVariables>(
     getDataTypeQuery,
     {
-      /** @todo reconsider caching. This is done for testing/demo purposes. */
-      fetchPolicy: "no-cache",
+      // Entity types are immutable, any request for an entityTypeId should always return the same value.
+      fetchPolicy: "cache-first",
```
Comment (benwerner01) [inline packages/hash/frontend/src/components/hooks/blockProtocolFunctions/ontology/useBlockProtocolGetDataType.ts:18]:
```suggestion
      // Data types are immutable, any request for an dataTypeId should always return the same value.
      fetchPolicy: "cache-first",
```

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts [hunk 2]
```diff
@@ -17,10 +17,7 @@ export const useBlockProtocolGetEntity = (): {
   const [getEntityTypeRootedSubgraphFn] = useLazyQuery<
     GetEntityTypeRootedSubgraphQuery,
     GetEntityTypeRootedSubgraphQueryVariables
-  >(getEntityTypeRootedSubgraphQuery, {
-    /** @todo reconsider caching. This is done for testing/demo purposes. */
-    fetchPolicy: "no-cache",
-  });
+  >(getEntityTypeRootedSubgraphQuery);
```
Comment (benwerner01) [inline packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts:20]: What is the default behaviour here now? Is it still `no-cache`?

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/ontology/useBlockProtocolGetLinkType.ts [hunk 3]
```diff
@@ -14,8 +14,8 @@ export const useBlockProtocolGetLinkType = (): {
   const [getFn] = useLazyQuery<GetLinkTypeQuery, GetLinkTypeQueryVariables>(
     getLinkTypeQuery,
     {
-      /** @todo reconsider caching. This is done for testing/demo purposes. */
-      fetchPolicy: "no-cache",
+      // Entity types are immutable, any request for an entityTypeId should always return the same value.
+      fetchPolicy: "cache-first",
```
Comment (benwerner01) [inline packages/hash/frontend/src/components/hooks/blockProtocolFunctions/ontology/useBlockProtocolGetLinkType.ts:18]:
```suggestion
      // Link types are immutable, any request for an linkTypeId should always return the same value.
      fetchPolicy: "cache-first",
```

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolUpdateEntity.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+      const { data: responseData } = await updateFn({
+        variables: {
+          entityId,
+          updatedProperties,
+        },
+      });
+
+      const { updatePersistedEntity: updatedEntity } = responseData ?? {
+        updatePersistedEntity: null,
+      };
```
Comment (benwerner01) [inline packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolUpdateEntity.ts:63]: Can't we just do this instead?
```suggestion
      const { updatePersistedEntity: updatedEntity } = responseData ?? {};
```

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts (hunk 2 above)
Comment (thehabbos007) [reply packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts:20]: Ah good point, perhaps we should be making use of the hook here instead? 
The default would be `cache-first` which is what we would want for any `get*Type` query as the IDs refer to immutable records

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolUpdateEntity.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+              code: "INVALID_INPUT",
+              message: "Error calling updatePersistedEntity",
+            },
+          ],
+        };
+      }
+
+      /**
+       * @todo: the EntityTypeRootedSubraph is not returned from the `updateEntity` mutation.
+       * May be addressed as part of https://app.asana.com/0/1200211978612931/1203089535761796/f or related work.
```
Comment (benwerner01) [inline packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolUpdateEntity.ts:78]:
```suggestion
       * @todo: the EntityTypeRootedSubgraph is not returned from the `updateEntity` mutation.
       * May be addressed as part of https://app.asana.com/0/1200211978612931/1203089535761796/f or related work.
```

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts (hunk 2 above)
Comment (thehabbos007) [reply packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolGetEntity.ts:20]: I think I do want to replace this with the hook, actually. It is temporary, but the code with the hook instead is way shorter

**Review by benwerner01 — APPROVED** [review]

**Pull request merged**